[flang] Initial debug info support for local variables.

abidh opened the pull request:
We need the information in the `DeclareOp` to generate debug information
for variables.  Currently, cg-rewrite removes the `DeclareOp`. As
`AddDebugInfo` runs after that, it cannot process the `DeclareOp`. My initial
plan was to make the `AddDebugInfo` pass run before the cg-rewrite
but that has few issues.
    
1. Initially I was thinking to use the memref op to carry the variable
attr. But as @tblah suggested in the #86939, it makes more sense to
carry that information on `DeclareOp`. It also makes it easy to handle it
in codegen and there is no special handling needed for arguments. For
this reason, we need to preserve the `DeclareOp` till the codegen.
    
2. Running earlier, we will miss the changes in passes that run between
cg-rewrite and codegen.
    
But not removing the DeclareOp in cg-rewrite has the issue that ShapeOp
remains and it causes errors during codegen. To solve this problem, I
convert DeclareOp to XDeclareOp in cg-rewrite instead of removing
it. This was mentioned as possible solution by @jeanPerier in
https://reviews.llvm.org/D136254
    
The conversion follows similar logic as used for other operators in that
file. The FortranAttr and CudaAttr are currently not converted but left
as TODO when the need arise.

Now `AddDebugInfo` pass can extracts information about local variables
from `XDeclareOp` and creates `DILocalVariableAttr`. These are attached
to `XDeclareOp` using `FusedLoc` approach. Codegen can use them to
create `DbgDeclareOp`.  I have added tests that checks the debug
information in mlir from and also in llvm ir.

Currently we only handle very limited types. Rest are given a place holder
type. The previous placeholder type was basic type with `DW_ATE_address`
encoding. When variables are added, it started causing assertions in
the llvm debug info generation logic for some types. It has been changes to
an interger type to prevent these issues until we handle those types properly.

abidh committed:
[flang] Preserve DeclareOp info in pre-cg-rewrite.

Currently, cg-rewrite removes the DeclareOp. As AddDebugInfo runs after
that, it cannot process the DeclareOp. My initial plan was to make the
AddDebugInfo pass run before the cg-rewrite but that has few issues.

1. Initially I was thinking to use the memref op to carry the variable
attr. But as @tblah suggested in the #86939, it makes more sense to
carry that information on DeclareOp. It also makes it easy to handle it
in codegen and there is no special handling needed for arguments. For
that, we need to preserve the DeclareOp till the codegen.

2. Running earlier, we will miss the changes in passes that run between
cg-rewrite and codegen.

But not removing the DeclareOp in cg-rewrite has the issue that ShapeOp
remains and it causes errors during codegen. To solve this problem, I
convert DeclareOp to XDeclareOp in cg-rewrite instead of removing
it. This was mentioned as possible solution by @jeanPerier in
https://reviews.llvm.org/D136254

The conversion follows similar logic as used for other operators in that
file. The FortranAttr and CudaAttr are currently not converted but left
as TODO when the need arise. A later commit will use the XDeclareOp to
extract the variable information.

abidh committed:
[flang] Add initial support for local variables.

This commit extracts information about local variables from XDeclareOp
and creates DILocalVariableAttr. These are attached to DeclareOp using
FusedLoc approach. Codegen can use them to create DbgDeclareOp.

I have added tests that checks the debug information in mlir from
and also in llvm ir.

abidh committed:
[flang] Use integer as a place holder type.

The previous placeholder type was basic type with DW_ATE_address
encoding. When variables are added, it started causing assertions in
the llvm debug info generation logic for some types. It has been changed
to an integer type.

tblah commented on flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:
```diff
@@ -45,13 +46,44 @@ namespace fir {
+  mlir::MLIRContext *context = &getContext();
+  mlir::OpBuilder builder(context);
+
+  bool isLocal = (declOp.getMemref().getDefiningOp() != nullptr);
```
I think block arguments would have no defining op. This could happen after structured control flow has been lowered to blocks and branches like in LLVM IR.

Off the top of my head, I think all of the source code variable declarations would end up in the entry block to the function and so wouldn't have this problem. It might be possible that declare operations for compiler temporaries come via block arguments (e.g. after control flow). I think these might then be mislabeled as dummy arguments, and end up incrementing `argNo`.

tblah reviewed (commented):
Thanks for your work on this!

Do we really need an XDeclare operation? It doesn't look like we would loose anything if we just allowed regular declare operations to persist until CodeGen.

jeanPerier commented on flang/lib/Optimizer/CodeGen/PreCGRewrite.cpp:
```diff
@@ -276,7 +276,29 @@ class DeclareOpConversion : public mlir::OpRewritePattern<fir::DeclareOp> {
+    }
+    // FIXME: Add FortranAttrs and CudaAttrs
+    auto xDeclOp = rewriter.create<fir::cg::XDeclareOp>(
+        loc, declareOp.getType(), declareOp.getMemref(), shapeOpers, shiftOpers,
```
Maybe this should only be created if there is a DILocalVariableAttr fused with the location of the DeclareOp.

Creating MLIR operations, especially with many arguments, has a compile time cost (both memory and time), so I would be inclined to start by only doing it when we know it is needed, which is easy here currently. That way, there is no extra compile time cost for filed compiled without -g.

Future use cases can enable this translation when they arise.

abidh replied:
This pre-cg-rewrite runs before `AddDebugInfo` so we would not have created `FusedLoc` by that time. I am thinking that we could run a cleanup pass before pre-cg-rewrite that removes the `DeclareOp` if debug info is disabled. Does that sound like the right approach to you?

jeanPerier replied:
Good point. I think adding a pass option "preserveDeclare" to pre-cg-rewrite  passwould be simpler and cheaper then. It can be set in createDefaultFIRCodeGenPassPipeline [here](https://github.com/llvm/llvm-project/blob/37f3661f4e8d180dc702ab42fa015bc3c21ceb51/flang/include/flang/Tools/CLOptions.inc#L308) given the debug level.

abidh replied:
Thanks for the suggestions. I have added the option in 2c566dd. Also updated the tests to check with this option turned off and on.

jeanPerier commented on flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:
```diff
@@ -45,13 +46,44 @@ namespace fir {
+  mlir::MLIRContext *context = &getContext();
+  mlir::OpBuilder builder(context);
+
+  bool isLocal = (declOp.getMemref().getDefiningOp() != nullptr);
```
I find `isLocal` a bit confusing here because global variables would also have "isLocal" set to true (since they are the result of a fir.address_of).

I guess reversing the logic and using "isDummyArgument" would described more what this is about. Looking for the existence of a defining op is also not solid because there are cases (e.g.) where there will be a little bit of argument "cooking" before the hlfir.declare (converts/unboxing). For instance in:

```
subroutine foo(c)
  character(10) :: c
end subroutine
```

Also, I am not sure what "argNo" ends-up being used for in the DWARF, but it is likely not accurate since hlfir.declare are created alphabetically (except when a variable "B" appears in the specification expression of another variable "A", in which case "B" declare is created before "A").
Does the position must be the Fortran argument position, or the one in the LLVM signature? These may differ since some Fortran function results may be passed by pointer via the first argument in the LLVM signature, and some Fortran arguments are split (The length of character arguments is passed after all Fortran arguments).

So we may need to add some info on the fir/hlfir declare to indicate something is a dummy argument, and what is its position.

Do you also know what should happen after inlining?

abidh replied:
The argNo value we pass will determine the order in which debugger will show the variables.

```
function fn1(c1, a1, b1) result (res1)
...
end function

(GDB) frame
#0  fn1 (c1=.FALSE., a1=1, b1=2.1000000000000001) at ../test1.f90:20

```
You were right that `DeclareOP` were in alphabetical order. I have now used `BlockArgument::getArgNumber` and it seems to return the correct order value. I am also using cast to BlockArgument to check if it is a dummy argument. But the problem of argument cooking in certain cases remain. I added a FIXME for that. Once I know all the cases where it can happen, it probably should not be too difficult to handle.

From inlining, you mean inlining that can happen before we process `DeclareOp` or after. If it happens after then I think it is up to that pass to adjust any metadata accordingly. Are there instances where inlining will happen before `AddDebugInfo` runs?

vzakhari replied:
MLIR inlining is not quite supported in FIR dialect, but you can experiment with some cases (some of them may not work) using `flang-new -O3 -mmlir -inline-all=true`, e.g. for this test:
```
subroutine test(x, y)
  real, target :: x, y
  x = y ! may alias
  call inner(x, y)
contains
  subroutine inner(x, y)
    real :: x, y
    x = y ! may not alias
  end subroutine inner
end subroutine test
```
I get this MLIR:
```
// -----// IR Dump After Inliner (inline) //----- //
module attributes {dlti.dl_spec = #dlti.dl_spec<#dlti.dl_entry<i1, dense<8> : vector<2xi64>>, #dlti.dl_entry<!llvm.ptr, dense<64> : vector<4xi64>>, #dlti.dl_entry<f80, dense<128> : vector<2xi64>>, #dlti.dl_entry<i32, dense<32> : vector<2xi64>>, #dlti.dl_entry<i16, dense<16> : vector<2xi64>>, #dlti.dl_entry<i8, dense<8> : vector<2xi64>>, #dlti.dl_entry<f128, dense<128> : vector<2xi64>>, #dlti.dl_entry<!llvm.ptr<271>, dense<32> : vector<4xi64>>, #dlti.dl_entry<f16, dense<16> : vector<2xi64>>, #dlti.dl_entry<!llvm.ptr<270>, dense<32> : vector<4xi64>>, #dlti.dl_entry<f64, dense<64> : vector<2xi64>>, #dlti.dl_entry<i128, dense<128> : vector<2xi64>>, #dlti.dl_entry<!llvm.ptr<272>, dense<64> : vector<4xi64>>, #dlti.dl_entry<i64, dense<64> : vector<2xi64>>, #dlti.dl_entry<"dlti.stack_alignment", 128 : i64>, #dlti.dl_entry<"dlti.endianness", "little">>, fir.defaultkind = "a1c4d8i4l4r4", fir.kindmap = "", fir.target_cpu = "x86-64", llvm.data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128", llvm.target_triple = "x86_64-unknown-linux-gnu"} {
  func.func @_QPtest(%arg0: !fir.ref<f32> {fir.bindc_name = "x", fir.target}, %arg1: !fir.ref<f32> {fir.bindc_name = "y", fir.target}) {
    %0 = fir.dummy_scope : !fir.dscope
    %1 = fir.declare %arg0 dummy_scope %0 {fortran_attrs = #fir.var_attrs<target>, uniq_name = "_QFtestEx"} : (!fir.ref<f32>, !fir.dscope) -> !fir.ref<f32>
    %2 = fir.declare %arg1 dummy_scope %0 {fortran_attrs = #fir.var_attrs<target>, uniq_name = "_QFtestEy"} : (!fir.ref<f32>, !fir.dscope) -> !fir.ref<f32>
    %3 = fir.load %2 : !fir.ref<f32>
    fir.store %3 to %1 : !fir.ref<f32>
    %4 = fir.dummy_scope : !fir.dscope
    %5 = fir.declare %1 dummy_scope %4 {uniq_name = "_QFtestFinnerEx"} : (!fir.ref<f32>, !fir.dscope) -> !fir.ref<f32>
    %6 = fir.declare %2 dummy_scope %4 {uniq_name = "_QFtestFinnerEy"} : (!fir.ref<f32>, !fir.dscope) -> !fir.ref<f32>
    %7 = fir.load %6 : !fir.ref<f32>
    fir.store %7 to %5 : !fir.ref<f32>
    return
  }
}
```

abidh replied:
Thanks. I looked at the IR in such case and it seems that function that gets inlined loses any `FuncOp`. Name in `DeclareOp` and `CallSiteLoc` may provide some clues but work will be required to retain enough information during inlining to enable `AddDebugInfo` to do its job. In any case, our first target should be to get the debug working with -O0.

jeanPerier reviewed (commented):
Thank you, this is a big step for debug info in flang!

abidh committed:
Handle review comments.

Following changes were done.

1. Discard variables if they are not of NameKind::VARIABLE kind.

2. Use casting to BlockArgument to check if it is a dummy argument and use getArgNumber() to get its number.

abidh commented on flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:
```diff
@@ -45,13 +46,44 @@ namespace fir {
+  bool isLocal = (declOp.getMemref().getDefiningOp() != nullptr);
+  auto tyAttr = typeGen.convertType(fir::unwrapRefType(declOp.getType()),
+                                    fileAttr, scopeAttr, declOp.getLoc());
+  auto result = fir::NameUniquer::deconstruct(declOp.getUniqName());
```
Thanks for the suggestion. Done now.

abidh committed:
[flang] Add `preserveDeclare` option to cg-rewrite pass.

When this option is true, DeclareOp is converted to XDeclareOp and this OP is used later for debug info generation. When it is false, DeclareOp is removed.

abidh commented:
Any further comments on this PR?

abidh committed:
Create DILocalVariable for local variables only.

The DeclareOp is also generated for module variable used in a function.
We reject them to avoid creating local variables for them.

tblah reviewed (approved):
This looks good to me. Thank you for working on this.

sscalpone commented:
@abidh @psteinfeld 

I am seeing an assertion that I believe is related to this commit.

File 90905.f90
```
module mm
private
public:: ff
contains
  subroutine ff()
  implicit none
    real(8):: x
    return
  end subroutine ff
end module
```
```
% flang-new -gline-tables-only -c 90905.f90
flang-new: /proj/build/llvm/Linux_x86_64/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:2303: virtual void llvm::DwarfDebug::endFunctionImpl(const llvm::MachineFunction*): Assertion `InfoHolder.getScopeVariables().empty()' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /proj/nv/llvm/Linux_x86_64/llvm-2887/bin/flang-new -fc1 -triple x86_64-unknown-linux-gnu -emit-obj -fcolor-diagnostics -mrelocation-model pic -pic-level 2 -pic-is-pie -target-cpu x86-64 -debug-info-kind=line-tables-only -resource-dir /proj/nv/llvm/Linux_x86_64/llvm-2887/bin/.. -mframe-pointer=all -o 90905.o -x f95-cpp-input 90905.f90
1.      Running pass 'Function Pass Manager' on module 'FIRModule'.
2.      Running pass 'X86 Assembly Printer' on function '@_QMmmPff'
 #0 0x0000556cf4b9d81b llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/proj/nv/llvm/Linux_x86_64/llvm-2887/bin/flang-new+0x259d81b)
 #1 0x0000556cf4b9ad74 SignalHandler(int) Signals.cpp:0:0
 #2 0x000014ddbe054db0 __restore_rt (/lib64/libc.so.6+0x54db0)
 #3 0x000014ddbe0a154c __pthread_kill_implementation (/lib64/libc.so.6+0xa154c)
 #4 0x000014ddbe054d06 gsignal (/lib64/libc.so.6+0x54d06)
 #5 0x000014ddbe0287f3 abort (/lib64/libc.so.6+0x287f3)
 #6 0x000014ddbe02871b _nl_load_domain.cold (/lib64/libc.so.6+0x2871b)
 #7 0x000014ddbe04dca6 (/lib64/libc.so.6+0x4dca6)
 #8 0x0000556cf6cf8357 llvm::DwarfDebug::endFunctionImpl(llvm::MachineFunction const*) (/proj/nv/llvm/Linux_x86_64/llvm-2887/bin/flang-new+0x46f8357)
 #9 0x0000556cf6d548a6 llvm::DebugHandlerBase::endFunction(llvm::MachineFunction const*) (/proj/nv/llvm/Linux_x86_64/llvm-2887/bin/flang-new+0x47548a6)
#10 0x0000556cf6cbbe91 llvm::AsmPrinter::emitFunctionBody() (/proj/nv/llvm/Linux_x86_64/llvm-2887/bin/flang-new+0x46bbe91)
#11 0x0000556cf46b6821 llvm::X86AsmPrinter::runOnMachineFunction(llvm::MachineFunction&) (/proj/nv/llvm/Linux_x86_64/llvm-2887/bin/flang-new+0x20b6821)
#12 0x0000556cf7b9d931 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (.part.0) MachineFunctionPass.cpp:0:0
#13 0x0000556cf96f92ab llvm::FPPassManager::runOnFunction(llvm::Function&) (/proj/nv/llvm/Linux_x86_64/llvm-2887/bin/flang-new+0x70f92ab)
#14 0x0000556cf96f94f1 llvm::FPPassManager::runOnModule(llvm::Module&) (/proj/nv/llvm/Linux_x86_64/llvm-2887/bin/flang-new+0x70f94f1)
#15 0x0000556cf96f9d95 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/proj/nv/llvm/Linux_x86_64/llvm-2887/bin/flang-new+0x70f9d95)
#16 0x0000556cf4ffee13 Fortran::frontend::CodeGenAction::executeAction() (/proj/nv/llvm/Linux_x86_64/llvm-2887/bin/flang-new+0x29fee13)
#17 0x0000556cf4bda07d Fortran::frontend::FrontendAction::execute() (/proj/nv/llvm/Linux_x86_64/llvm-2887/bin/flang-new+0x25da07d)
#18 0x0000556cf4bc9deb Fortran::frontend::CompilerInstance::executeAction(Fortran::frontend::FrontendAction&) (/proj/nv/llvm/Linux_x86_64/llvm-2887/bin/flang-new+0x25c9deb)
#19 0x0000556cf4be0045 Fortran::frontend::executeCompilerInvocation(Fortran::frontend::CompilerInstance*) (/proj/nv/llvm/Linux_x86_64/llvm-2887/bin/flang-new+0x25e0045)
#20 0x0000556cf45ea828 fc1_main(llvm::ArrayRef<char const*>, char const*) (/proj/nv/llvm/Linux_x86_64/llvm-2887/bin/flang-new+0x1fea828)
#21 0x0000556cf451ba8e main (/proj/nv/llvm/Linux_x86_64/llvm-2887/bin/flang-new+0x1f1ba8e)
#22 0x000014ddbe03feb0 __libc_start_call_main (/lib64/libc.so.6+0x3feb0)
#23 0x000014ddbe03ff60 __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x3ff60)
#24 0x0000556cf45e954e _start (/proj/nv/llvm/Linux_x86_64/llvm-2887/bin/flang-new+0x1fe954e)
flang-new: error: unable to execute command: Aborted (core dumped)
flang-new: error: flang frontend command failed due to signal (use -v to see invocation)
flang-new version 19.0.0git (/proj/build/llvm/Linux_x86_64/clang dcf3102be8458fe7588f9d11315beddfca4323b0)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /proj/nv/llvm/Linux_x86_64/llvm-2887/bin
Build config: +assertions
flang-new: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
flang-new: note: diagnostic msg: /tmp/90905-da4491
flang-new: note: diagnostic msg: /tmp/90905-da4491.sh
flang-new: note: diagnostic msg:

********************
```

abidh commented:
The problem was that variables were being generated even when user just asked for line-tables. I have fixed it now. I will re-land it after testing.

psteinfeld commented:
> The problem was that variables were being generated even when user just asked for line-tables. I have fixed it now. I will re-land it after testing.

Thanks, @abidh.  Let me know if you want me to test something.
